### ISSUE TITLE
Two emails sent in call to /account/login whenever account is unverified

### ISSUE THREAD

**Issue by shane-tomlinson** [mozilla/fxa-auth-server#1314]
STR

~~1. Sign up with a new email, do not verify.
1. Call /account/login with new email, both the confirm signin and confirm signup emails are sent.~~

This can be seen using the content server and auth-mailer logs

1. Sign up, do not verify,
2. Sign in with account from step 1, look at the auth-mailer logs

![screen shot 2016-06-29 at 16 34 05](https://cloud.githubusercontent.com/assets/848085/16458579/736ad26c-3e17-11e6-92fc-16eb3a099ce8.png)

**Comment by shane-tomlinson** [comment]
Ref https://github.com/mozilla/fxa-content-server/issues/3890

**Comment by shane-tomlinson** [comment]
@vbudhram - I believe [this test is incorrect](https://github.com/mozilla/fxa-auth-server/blob/282271b9554cfd28c86c3ba1e37ccd02c7aee7d0/test/remote/account_signin_verification_tests.js#L230). Since the user is signing in with an unverified account, the subject should *still* be `Verify your Firefox Account` as on line 213. The premise of the tests between [lines 263 and 344](https://github.com/mozilla/fxa-auth-server/blob/282271b9554cfd28c86c3ba1e37ccd02c7aee7d0/test/remote/account_signin_verification_tests.js#L263:L344) seems incorrect as well. The user should not be able to verify a session only at this point, they should only be able to verify their account. By verifying the account, their session should be automatically verified.

**Comment by shane-tomlinson** [comment]
This patch seems to fix the behavior, but not the tests:

```diff
diff --git a/lib/routes/utils/request_helper.js b/lib/routes/utils/request_helper.js
index 146d3f8..d7b17f8 100644
--- a/lib/routes/utils/request_helper.js
+++ b/lib/routes/utils/request_helper.js
@@ -22,6 +22,11 @@ function wantsKeys (request) {
  */
 function shouldEnableSigninConfirmation(account, config, request) {
 
+  if (! account.emailVerified) {
+    return false
+  }
+
+
   var confirmLogin = config.signinConfirmation && config.signinConfirmation.enabled
   if (!confirmLogin) {
     return false
```

**Comment by vbudhram** [comment]
Dug into this a little more and found that even with sign-in confirmation disabled, you would still get two emails when signing in with sync, `New sign-in to Firefox` and `Verify your Firefox Account`. On the content-server side, it checks to see if [account is verified](https://github.com/mozilla/fxa-content-server/blob/master/app/scripts/models/user.js#L360) and if not, requests resend_code.

Previously, the `/account/login` endpoint didn't send any emails so it made sense to have content-server request to resend codes. However, this endpoint now sends two emails, does it still make sense to have content-server request codes?

In either case, I do think that unverified users should always get the `Verify your Firefox` email when logging in.

**Comment by shane-tomlinson** [comment]
> `New sign-in to Firefox` and `Verify your Firefox Account`

This seems incorrect to me. In this case, only sending the `Verify your Firefox Account` makes sense.

> Previously, the /account/login endpoint didn't send any emails so it made sense to have content-server request to resend codes. However, this endpoint now sends two emails, does it still make sense to have content-server request to resend verify codes on new login?

It probably does. If we have the auth server send the email w/o the extra call, we'll need to ensure the content server can handle both the new and old auth server, and that the auth server can handle both the old and new content server. That might require a version bump, or sending a flag from the content server to the auth server to indicate "send unverified user emails, if possible", and the auth server would have to return a bit that says "email already sent, don't worry about it." I'm not against it for any ideological reasons, but might be more trouble than it's worth for now.

> In either case, I do think that unverified users should always get the Verify your Firefox email when logging in.

Yes, totally.

**Comment by rfk** [comment]
>  >  New sign-in to Firefox and Verify your Firefox Account
> This seems incorrect to me. In this case, only sending the Verify your Firefox Account makes sense.

I lightly disagree - the "new signin to firefox" email also serves as a security check and will soon include geo information.  If we get rid of it for some percentage of signins, we may need to replicate that additional security information into whatever email we do send.

Maybe we *should* in fact add that information to the "verify your account" email, at least for sync logins?

**Comment by vbudhram** [comment]
> Maybe we should in fact add that information to the "verify your account" email, at least for sync logins?

I think that is a fair proposal. 

Currently, as an unverified account performing a login, you will get either a `New sign-in to Firefox` or `Confirm sign-in to Firefox`. Should this patch also not send the `New sign-in to Firefox`? My initial impression is no, since it seems strange to send two emails.

**Comment by rfk** [comment]
If the user clicks through to verify their account when connecting the second device, does that cause both sessions to become confirmed and start syncing?  I feel like we added some logic so that it only verified the session from which the email originated, but want to make sure.

Broadly, I agree we should only send a single email, but we have to make sure it's conveying all the right information.

**Comment by seanmonstar** [comment]
I'd worry that a user could ignore a "hey you, verify your email!", when it might have information *inside* it saying "there was a new sign in across the world, does that sound right?".

They could even just see the email, think it's a reminder, and verify their email, but never notice that an attacker logged in as well.

tl;dr - I'd rather there be 2 emails, unless the UX can make it abundantly clear.

**Comment by rfk** [comment]
> They could even just see the email, think it's a reminder, 

Right, this is the part that most concerns me as well, especially since we are actually going to send them reminder emails.

**Comment by vbudhram** [comment]
> only verified the session from which the email originated

That is correct, the email will verify the session that triggered the verification. Doubled checked this by logging in from multiple browsers.

**Comment by shane-tomlinson** [comment]
> If the user clicks through to verify their account when connecting the second device, does that cause both sessions to become confirmed and start syncing? I feel like we added some logic so that it only verified the session from which the email originated, but want to make sure.

It depends. If you sign up, never verify, then sign in, then verify, yes, both sessions become confirmed and start syncing. I just tested this locally. 

If you sign up, never verify, then re-sign-up, verify, then no, it'll create a new account. I just tested this locally too.

> I'd worry that a user could ignore a "hey you, verify your email!", when it might have information inside it saying "there was a new sign in across the world, does that sound right?".

This leads me to believe if a user signs in with an unverified account that already exists, we should send a *different* email to either of the existing emails that is a cross between the "verify this signup" and "new sign in". "Hey, you previously signed up with the same email and password, but never verified. If that computer is still connected to Sync, it'll start syncing with you. Here is where you are signing up from now, and here is where you signed up from before." As a stopgap we could display the geo information in the signup email.

> tl;dr - I'd rather there be 2 emails, unless the UX can make it abundantly clear.

Testing locally, I found 2 emails confusing. 2 emails makes no sense when they are triggered by the same action. I lean heavily towards making the UX abundantly clear in a single email.

Finally, from my own lengthy response, I'm afraid starting to wandering into bike shed territory and let the idea of perfect get in the way of good enough to get out the door.

**Comment by rfk** [comment]
Hmm, @vbudhram @shane-tomlinson we have conflicting reports here:

> That is correct, the email will verify the session that triggered the verification.
> Doubled checked this by logging in from multiple browsers.

versus

> It depends. If you sign up, never verify, then sign in, then verify, yes,
> both sessions become confirmed and start syncing. I just tested this locally. 

I wonder if the cause is e.g. which of the multiple emails you end up clicking through.  Worth digging in deeper here to see if we can all get to a shared understanding of what the existing behaviour should be.

**Comment by shane-tomlinson** [comment]
> Hmm, @vbudhram @shane-tomlinson we have conflicting reports here:

No conflict, we outlined two different flows. @vbudhram outlined a verified user going through the signin flow in two browsers. I outlined sign up (do not verify), then sign in (with the unverified email/pw combo) in a 2nd browser, then verify the email sent from the signin.

In the flow I outlined, the user receives the exact same sign up verification email twice. @seanmonstar pointed out that both browsers will start syncing with each other as soon as the user clicks the verification link in either email, and we should change the email UX so that a user understands the repercussions. I agree with this. Either that, or the first device should *not* start Syncing until the user actually confirms the email sent for that device.

**Comment by rfk** [comment]
> Either that, or the first device should not start Syncing until the user actually confirms
> the email sent for that device.

I completely agree with the concern about both browsers starting to sync, it's just that I thought the above was the behavior we already had - the only session that gets verified, is the one that caused the email to be sent.  But clearly it's not the case.

> we are starting to wander into bike shed territory and let the idea of perfect get 
> in the way of good enough to get out the door.

Good point, and I guess there are two separate questions here - what's the incremental behaviour change we want to see for when this first goes live, and what's the ideal behaviour we might like to refactor our way towards?
